nexus/db-queries/src/db/datastore/ssh_key.rs

```diff
@@ -21,11 +21,200 @@ use diesel::prelude::*;
+
+        // If a key isn't present in the result that was present in the input that means it either
+        // doesn't exist or isn't owned by the user. Either way we want to give a specific lookup error
+        // for at least the first result. It would be nice to include an aggregate error with all the missing
+        // keys.
+        // TODO: Optimize the performance of this
+        for key in keys.iter() {
+            match key {
+                NameOrId::Name(name) => {
+                    if !result
+                        .iter()
+                        .any(|(_, n)| n.clone() == name.clone().into())
+                    {
+                        return Err(Error::ObjectNotFound {
+                            type_name: ResourceType::SshKey,
+                            lookup_type: LookupType::ByName(name.to_string()),
+                        });
+                    }
+                }
+                NameOrId::Id(id) => {
+                    if !result.iter().any(|(i, _)| i == id) {
+                        return Err(Error::ObjectNotFound {
+                            type_name: ResourceType::SshKey,
+                            lookup_type: LookupType::ById(*id),
+                        });
+                    }
+                }
+            }
+        }
+
+        return Ok(result.iter().map(|&(id, _)| id).collect());
+    }
+
+    /// Given a list of IDs for SSH public keys, fetches the keys that belong to
+    /// the user and aren't deleted. Does not fail if keys are missing.
+    pub async fn ssh_keys_batch_fetch(
+        &self,
+        opctx: &OpContext,
+        authz_user: &authz::SiloUser,
+        keys: &Vec<Uuid>,
+    ) -> ListResultVec<SshKey> {
+        opctx.authorize(authz::Action::ListChildren, authz_user).await?;
+
+        use db::schema::ssh_key::dsl;
+        dsl::ssh_key
+            .filter(dsl::id.eq_any(keys.to_owned()))
+            .filter(dsl::silo_user_id.eq(authz_user.id()))
+            .filter(dsl::time_deleted.is_null())
+            .select(SshKey::as_select())
+            .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    // Associate a list of SSH public keys with an instance. This happens
+    // during the instance create saga and does not fail if the the ssh keys
+    // have been deleted as a race condition.
+    pub async fn ssh_keys_batch_assign(
+        &self,
+        opctx: &OpContext,
+        authz_user: &authz::SiloUser,
+        instance_id: Uuid,
+        keys: &Option<Vec<Uuid>>,
+    ) -> UpdateResult<()> {
+        opctx.authorize(authz::Action::ListChildren, authz_user).await?;
+
+        let instance_ssh_keys: Vec<db::model::InstanceSshKey> = match keys {
+            // If the keys are None, use the fallback behavior of assigning all the users keys
+            None => {
```

Comment:
Would this not confuse users thinking that `None` means add no keys? Perhaps we may want to revisit why the default behaviour is to add all keys.

There are a few reasons we may want to consider this default behaviour:
- From a security standpoint, people may have forgotten to remove keys from their user account that have been compromised.
- Additionally, users may not be aware that `None` means all, it doesn't feel particularly intuitive (often people don't read the docs!). They may end up adding keys they did not intend to add, and be unaware of it if the instance was not created through the console.
- Do we expect our customers to particularly _want_ to have all keys automatically added to the instances they create?

I think it'd also be good to have some input from @ahl and/or @askfongjojo  on this

Reply:
Also, if users have more than 10 keys saved in their user account, every attempt at creating an instance with the default behaviour of adding all keys would result in an error. 

This means that someone with more than 10 keys would have to set which keys they want added every single time to avoid a failure.

Reply:
Heh, yeah, I've had a lot of conversations about this. @iliana and I [discussed](https://matrix.to/#/!oSMXTSrgambaNFEemN:oxide.computer/$SN8Tm7fxeQwbmlZhkrIua6pRWIwrWK8OEpnvl5v3QwQ?via=oxide.computer&via=unix.house&via=matrix.org) this in chat. This implementation was xer original suggestion. 

There's a few things to consider here. 
1. Should this be a required field? If it's required then it's one extra thing folks _have_ to specify when creating an instance. Originally I was worried about the repercussions of this decision (and others like it) on the CLI. That's mostly moot given that most args have to be provided via a json file. If it's _not_ required, I think the only sane default is to send them all (as we do today). Having the default mean none get sent would lead to workflow situations where users forget and then lock themselves out (which is worse in my mind than having to go and remove some keys manually after the fact)
2. Should we even have the option for users to send all keys to the instance? I feel like _yes_ is probably the answer here because I suspect most users will have few keys. That's also another reason I've leaned towards this behavior as the default. If we didn't preserve this option then users would _always_ have to specify keys by name/id in the API (lest they can't SSH into the instance) and that feels like a pretty big DX regression. That said, I would appreciate more feedback from folks on this.

So, yeah, an explicit enum having `AllFromProfile | Only[keys] | NoKeys` is certainly an option I can implement in this PR if that's the consensus. I do see the value of it, especially around _forcing_ users to make a choice. If the choice that they're forced to make is often the same as the default of this PR though, that feels not great.

Reply:
I really don't like breaking old clients. Requiring an explicit enum will be a flag day for any existing code. This is, I think, an extremely good candidate for being our first version bump on an API. We did give ourselves a number of backstops to be able to declare a new version of the API, but we haven't used any of them yet, so that sort of explodes the work needed to be done here.

If we want to land this in the immediate term, we should keep the implicit behavior to use all of the user's keys if the field is not present (i.e. `None`). But ideally we have a clearer v2 either way.

Reply:
Thanks for the comments both! These are my thoughts:

> Should this be a required field? If it's required then it's one extra thing folks have to specify when creating an instance. Originally I was worried about the repercussions of this decision (and others like it) on the CLI. That's mostly moot given that most args have to be provided via a json file. If it's not required, I think the only sane default is to send them all (as we do today). Having the default mean none get sent would lead to workflow situations where users forget and then lock themselves out (which is worse in my mind than having to go and remove some keys manually after the fact)

To be perfectly honest, I don't see this being a required field as a problem. IMO it's better to be explicit about what you're doing than get surprises down the line. Alternatively, if we don't want this to be a required field, we could also have a concept of a "default" SSH key which gets added to any instance where SSH keys haven't been defined. WDYT?

We also need to think of the use case where new users are migrating their workloads from their previous vendor. For example: a user that didn't have support for adding SSH keys at create time (*cough* VMware *cough*) will have had to either have a user data init script that creates authentication, or have an image with a predefined user, and add additional SSH keys later. In this scenario, users may want to initially conserve their current workflows as part of a big lift and shift operation. These migrations are no small thing. and we really need to think about how to make it as seamless as possible for our customers to migrate over without unintended consequences.

Automatically adding every SSH key the user has saved feels quite nuclear, and can lead to easily giving unintended access. Especially if you're in the type or role where your job is to set up infrastructure for others to use and have saved other people's public keys in your account.

FWIW, I don't see being locked out of your instance as a necessarily terrible thing. This is something the user would automatically notice and would only need to delete and recreate their instance.

> Should we even have the option for users to send all keys to the instance? I feel like yes is probably the answer here because I suspect most users will have few keys. That's also another reason I've leaned towards this behavior as the default. If we didn't preserve this option then users would always have to specify keys by name/id in the API (lest they can't SSH into the instance) and that feels like a pretty big DX regression. That said, I would appreciate more feedback from folks on this.

I wouldn't necessarily say most users would have a small amount of keys, it really depends. Also, users that manage infrastructure will have ssh keys of others. So it's not just how many keys are saved, but also to whom those keys are giving access to. Frankly, at this point I don't think we have enough information about what most users will be doing. To me the big questions for now are:

- Does anyone really need to have every key they have saved added to their instance on create?
- What is more annoying? Adding a flag to define ssh keys on instance create, or having to go into an instance and remove SSH keys that you realised you really shouldn't have in there later down the track.
- Which repercussions are worse? Unintentionally giving access to keys that shouldn't have this access (could be compromised keys the user forgot to remove or people who shouldn't have access), or having to recreate an instance because you forgot to define a key and got locked out (this could be mitigated by having a "default" ssh key).

Additionally, by automatically adding all keys we'd be making it harder for our users to follow the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege) 

> I really don't like breaking old clients. Requiring an explicit enum will be a flag day for any existing code. This is, I think, an extremely good candidate for being our first version bump on an API. We did give ourselves a number of backstops to be able to declare a new version of the API, but we haven't used any of them yet (as far as I'm aware?), so that sort of explodes the work needed to be done here.

As for the concern about breaking APIs, I think the consensus is that now is the moment to be breaking the API to make sure we get it right. Posting a recent comment here -> https://github.com/oxidecomputer/omicron/pull/4694#discussion_r1442402379

Reply:
I agree on breaking the API and including no keys if none are defined.

However making it required might be a poor user experience in the CLI. If the user doesn't want to include keys, they shouldn't need that argument.

Reply:
I appreciate all the discussion here. @karencfv you bring up a lot of great points and I feel like there's just ultimately some more product discovery we need to do here. 

I have higher priorities than this PR and I've already sank too much time into it. I'm just going to create an issue to encapsulate the context here and we can solve this in a follow up. The default behavior as implemented is the same as what we have today so it shouldn't be disruptive to include in this release while also giving us the time we need to figure out how we want this to be.

Reply:
Here's the follow up issue to tackle this work: https://github.com/oxidecomputer/omicron/issues/4919